lib/pbench/server/api/resources/query_apis/metadata_index/table_of_content.py

```diff
@@ -0,0 +1,95 @@
+    JSON,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.api.resources.query_apis import CONTEXT
+from pbench.server.api.resources.query_apis.metadata_index import RunIdBase
+
+
+class TableOfContent(RunIdBase):
+    """
+    Get the names of controllers within a date range.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("run_id", ParamType.STRING, required=True),
+                Parameter("parent", ParamType.STRING, required=True),
+            ),
+        )
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        parent = json_data.get("parent")
+        run_id = context["run_id"]
+        dataset = context["dataset"]
+
+        self.logger.info(
+            "Discover Table of Content for run_id {} and directory {}", run_id, parent,
+        )
+
+        # Retrieve the ES indices that belong to this run_id from the metadata
+        # table
+
+        indices = self.get_index(dataset, "run-toc")
+        if not indices:
+            self.logger.debug(
+                f"Found no indices matching the prefix run-toc"
+                f"for a dataset {dataset!r}"
+            )
+            abort(HTTPStatus.NOT_FOUND, message="Found no matching indices")
+
+        return {
+            "path": f"/{indices}/_search",
+            "kwargs": {
+                "json": {
+                    "size": 100,
+                    "query": {
+                        "bool": {
+                            "filter": [
+                                {"match": {"run_data_parent": run_id}},
+                                {
+                                    "dis_max": {
+                                        "queries": [
+                                            {"term": {"directory": parent}},
+                                            {"regexp": {"directory": f"{parent}[^/]+"}},
+                                        ]
+                                    }
+                                },
+                            ],
+                            "must_not": {
+                                "regexp": {"directory": f"{parent}[^/]+/[^/]+"}
+                            },
+                        }
+                    },
+                }
+            },
+        }
+
+    def postprocess(self, es_json: JSON, context: CONTEXT) -> JSON:
+        dir_list = []
+        file_list = []
+        for toc in es_json["hits"]["hits"]:
+            directory = toc["_source"]["directory"]
+            parent = toc["_source"]["parent"]
+            direct = directory.rsplit("/", 1)[0]
+            try:
+                if (
+                    parent != directory
+                ):  # This condition is especially when parent and directory = '/'
+                    t = toc["_source"]["name"]
+                    dir_list.append(t)
```

Comment:
I don't think the local variable buys us anything here -- if it will all fit on one line, then I would just reference `toc` directly in the `append()` call.